### ISSUE TITLE
"Virez-moi le pyramide des tests" - Vraiment?

### ISSUE THREAD

**Issue by martinsson** [42skillz/kata-TrainReservation#4]
La valeur de la pyramide des tests est dans mon interprétation et utilisation une réponse aux problématiques suivantes
- rapidité
- résout le pb des comportements combinatoires
- tests plus faciles du happy-path
- coût de maintenance des tests d'une grosse appli

Tu as déjà répondu à comment tu préserves la valeur du coût de maintenance dans le 1er article. Les autres je suis curieux. Je suis conscient que ces potentiels problèmes ne sont pas importantes dans des petites applications, donc si la réponse est "pas pyramide, sauf grosse monolithe" alors ca me va 👍

**Comment by tpierrain** [comment]
Merci pour tes retours Johan, j'aborderai tous ces sujets dans les épisodes qui vont suivre. Pour info, je viens de poster le second épisode aujourd'hui : https://github.com/42skillz/kata-TrainReservation/blob/master/RidMeOfThoseTestingPyramids-ep2.md

**Comment by martinsson** [comment]
Bonjour Thomas 
Je crois que je n'ai pas eu la réponse à ma question finalement.
J'ai l'impression que tu incites aux tests haut niveau sans compter. Sans parler de comment tu adresses les lenteurs qui arrivent avec le grossissement des projets. Ça me laisse dubitatif.

**Comment by tpierrain** [comment]
Salut Johan,
Si tu n'as pas eu de réponse, c'est par manque de temps (et d'organisation aussi surement), j'en suis désolé. Et ce n'est surtout pas pour éviter ta question légitime ;-)

Bon. Etant donné que je stub tous les trucs techniques (middleware, db, etc), mes tests d'acceptation, malgré leur coarse-grained (c.ad. tout le systeme sauf les I/Os) ne sont pas lents. En résumé: pas d'I/O, pas de lenteur. Mais ceci ne peut également fonctionner que si tu prends l'habitude d'écrire des tests rapides (mon impatience à attendre et mon addiction au feedback court - type NCrunch sont suffisant pour me motiver ;-)

Pour les problèmes de comportement combinatoire, j'essaie de test-caser certains de ces tests d'acceptation, voire de me faire une DSL de test pour que leur écriture soit la plus aisée possible, mais je dois avouer que certains cas sont tout de même couverts par des tests unitaires (fine-grained). Attention, je dis que j'en fait moins qu'avant, je ne dis pas que je n'en fait plus.

Cela réponds t-il à ta question ?

**Comment by martinsson** [comment]
Oui. Si test tests haut niveau tournent en permanence avec NCrunch (ou l'équivalent), qu'il arrive à faire la sélection des tests, que cela va très vite, qu'il n'y aucun élément aléatoire, qu'on ne dépend d'aucune service ni de base de données même in-memory. Ni que tu sois pénalisé par l'initialisation d'un framework de DI. Que les échecs soient facile à débouguer car de toutes façons tu n'as jamais plus de une minute (au pif) de travail dans le pipe entre deux tests verts. Ou bien que ces impacts soient suffisamment négligeables. Alors effectivement quel intérêt d'une pyramide des tests?

Je ne sais pas si mon analyse te correspond, mais dans le hypothèse que oui, qu'en penses tu de cette formulation.

_Faisons en sorte de rendre les tests métier/haut niveau faciles à maintenir. De cette façon nous pouvons en avoir bcp plus. L'avantage est une meilleur refactorabilité de l'application et des tests plus oriénté métier sans contrepartie_

En gros ta pyramide n'est plus un /\ mais prend la forme d'un ( ), avec la plupart des tests au niveau intermédiaire. Ca n'a beau plus avoir la forme d'une pyramide ta réflexion semble identique à celui qui a mené à trouver la pyramide des tests.

**Comment by tpierrain** [comment]
Je te confirme toutes tes hypothèses. Mais quitte a prendre une métaphore, je préfère largement celle du losange oui.

La pyramide fait croire aux gens qu’il faut faire plus de tests unitaires que tous les autres types de tests mais encore une fois, 90% des devs comprennent « test unitaire » par petit test d’une toute petite partie de mon soft.

Il n’y a que certains de mes amis « puristes » (qui font partie des 10% restant) pour me dire que ces tests d’acceptations sont également des tests « unitaires ». Il n’y a pas de problème -je le sais moi aussi- mais ce qui me gêne dans cette formulation « unitaire » à taille variable, c’est son coté confus et piégeur pour l’immense partie des gens.
